This Python script visualizes Sentinel-2 13-band data to differentiate urban (red), vegetation (green), and water (blue) areas using Linear Discriminant Analysis (LDA). It was trained on 80 Sentinel-2 L1C images (2015–2019) from La Rioja, Spain, with land use classifications from the SIOSE dataset.

Each RGB channel is generated through a distinct LDA transformation, maximizing information for urban, crop, and water-related classes. A multiclass classifier aligns non-target classes to zero, with absolute values highlighting the target classes.

The script produces intuitive maps: urban areas appear red, vegetation green, and water blue, though some misclassification (e.g., red crops, non-blue rivers) can occur.

In [1]:
def calculate_values(B02, B03, B04, B08, B8A, B11, B12):
    result = [
        abs(0.4809 * (
            ((((B02 * 2.5) - 0.3329) * -9.4425) +
             (((B03 * 2.5) - 0.3182) * 2.1846) +
             (((B04 * 2.5) - 0.3380) * 2.5333) +
             (((B11 * 2.5) - 0.5644) * 9.9256) +
             (((B12 * 2.5) - 0.4216) * -13.6911))
            - 0.4766
        )),
        abs(0.3275 * (
            ((((B02 * 2.5) - 0.2844) * 13.3644) +
             (((B03 * 2.5) - 0.2736) * -6.6588) +
             (((B04 * 2.5) - 0.2750) * -1.1994) +
             (((B08 * 2.5) - 0.5972) * -0.2090) +
             (((B8A * 2.5) - 0.6648) * 5.1630) +
             (((B11 * 2.5) - 0.5651) * -7.2183))
            + 0.0463
        )),
        abs(0.2361 * (
            ((((B03 * 2.5) - 0.2429) * 21.8759) +
             (((B04 * 2.5) - 0.2321) * -6.0679) +
             (((B08 * 2.5) - 0.4371) * -3.0608) +
             (((B11 * 2.5) - 0.4146) * -4.4420))
            + 0.2061
        ))
    ]
    return result


In [2]:
# Example inputs
B02, B03, B04, B08, B8A, B11, B12 = 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7
result = calculate_values(B02, B03, B04, B08, B8A, B11, B12)
print(result)

[3.4402637424750004, 2.063868631125, 0.9105061073039998]


The LDA-based visualization technique can be useful for your flooding alert app by enabling the classification of terrain types from satellite imagery, helping to detect flood-prone areas. By using Sentinel-2 imagery, you can classify water bodies, vegetation, and urban areas, which is critical for identifying abnormal water expansions and flood impacts. The app could display real-time maps, with blue indicating flooded zones, while green highlights affected vegetation and red marks urban areas. This integration would allow your app to monitor flooding events, prioritize evacuation, and improve disaster response. Additionally, historical flood data from Sentinel-2 could enhance your app’s predictive capabilities, providing valuable insights for future events.

The result [3.4402637424750004, 2.063868631125, 0.9105061073039998] represents the output of the three LDA transformations applied to the Sentinel-2 data for the urban, vegetation, and water classification, respectively. Here's how to interpret each value:

3.4402637424750004:
This value corresponds to the LDA-transformed result for urban areas (red channel). A higher value indicates that the pixel belongs more strongly to the urban class, or an area with characteristics similar to urban zones.

2.063868631125:
This value represents the LDA-transformed result for vegetation (green channel). It suggests how closely a pixel corresponds to a vegetation class. Higher values indicate a stronger likelihood of the area being vegetation, such as crops or forests.

0.9105061073039998:
This value corresponds to the LDA-transformed result for water areas (blue channel). A higher value indicates a stronger classification as a water body, such as a river, lake, or flooded area.

In summary:
The values represent the degree of association of a specific pixel to urban, vegetation, and water classes, with higher values showing a stronger match to that class. For example, if this result came from a specific pixel in a satellite image, the area is most strongly classified as urban, followed by vegetation, with the least association to water.